composer/trainer/trainer.py

```diff
@@ -1204,35 +1226,73 @@ def _get_autoresume_checkpoint(
-                    pass
-        # Require all ranks to have local checkpoint if we wish to restore from it
-        latest_checkpoint_exists = self._device.tensor_to_device(
-            torch.tensor([os.path.exists(latest_checkpoint_path)], dtype=torch.uint8))
-        dist.all_reduce(latest_checkpoint_exists, reduce_operation='MIN')
-        # If latest checkpoint is saved locally, change load_path to it
-        if int(latest_checkpoint_exists.item()) == 1:
+            self._try_checkpoint_download(latest_checkpoint_path, save_latest_remote_file_name, loggers,
+                                          load_progress_bar)
+
+        if self.deepspeed_enabled:
+            # Require all ranks to have their own local checkpoint if we wish to restore from it for deepspeed
+            latest_checkpoint_exists_on_all_ranks = self._device.tensor_to_device(
+                torch.tensor([os.path.exists(latest_checkpoint_path)], dtype=torch.uint8))
+            dist.all_reduce(latest_checkpoint_exists_on_all_ranks, reduce_operation='MIN')
+
+            if int(latest_checkpoint_exists_on_all_ranks.item()) == 0:
+                raise RuntimeError('DeepSpeed was enabled, but checkpoints were not found for all ranks')
+
+            return latest_checkpoint_path
+        else:
+            # The checkpoint must at least exist for rank zero
+            latest_checkpoint_exists_on_rank_zero = self._device.tensor_to_device(
+                torch.tensor([os.path.exists(latest_checkpoint_path) and dist.get_global_rank() == 0],
+                             dtype=torch.uint8))
+            dist.all_reduce(latest_checkpoint_exists_on_rank_zero, reduce_operation='MAX')
+
+            if int(latest_checkpoint_exists_on_rank_zero.item()) == 0:
+                return None
+
+            # broadcast the local checkpoint path to all ranks
+            latest_checkpoint_path_list = [os.path.abspath(latest_checkpoint_path)]
+            dist.broadcast_object_list(latest_checkpoint_path_list, src=0)
+            latest_checkpoint_path = latest_checkpoint_path_list[0]
+
+            # broadcast the remote checkpoint path to all ranks
+            save_latest_remote_file_name_list = [save_latest_remote_file_name]
+            dist.broadcast_object_list(save_latest_remote_file_name_list, src=0)
+            save_latest_remote_file_name = save_latest_remote_file_name_list[0]
+
+            # download the checkpoint on local rank 0 of all nodes
+            if dist.get_local_rank() == 0 and not os.path.exists(latest_checkpoint_path):
+                log.debug(f'Attempting to download the checkpoint {save_latest_remote_file_name} on to all nodes')
+                os.makedirs(save_folder, exist_ok=True)
+                self._try_checkpoint_download(latest_checkpoint_path, save_latest_remote_file_name, loggers,
+                                              load_progress_bar)
+            dist.barrier()
```

Comment:
I don't think this barrier is necessary because of the reduce two lines down

Reply:
discussed offline, barrier is necessary so that `os.path.exists` is called after local rank 0 has finished downloading the checkpoint.